### PR TITLE
safe copy for C++

### PR DESCRIPTION
**Pull request by Baobaobear**

### CODE REVIEW

**Comment by Baobaobear** [comment]
Travis-ci got something wrong, need to restart build

File: sort.h [hunk 1]
```diff
@@ -797,6 +807,34 @@ void BITONIC_SORT(SORT_TYPE *dst, const size_t size) {
   }
 }
 
+#if SORT_SAFE_CPY
+
+void SORT_TYPE_CPY(SORT_TYPE *dst, SORT_TYPE *src, const size_t size) {
+  for (size_t i = 0; i < size; ++i) {
```
Comment (swenson) [inline sort.h]: You'll need to declare variables at the top of the function for C89.

File: sort.h [hunk 2]
```diff
@@ -19,6 +19,14 @@
 #define SORT_CMP(x, y)  ((x) < (y) ? -1 : ((y) < (x) ? 1 : 0))
 #endif
 
+#ifndef SORT_SAFE_CPY
```
Comment (swenson) [inline sort.h]: I think I would like to make this user-controlled, and the default be `#define SORT_SAFE_CPY 0`.

The assumption of this library is that you have a contiguous array C types, so standard memmove and memcpy should be fine.

I think if users are relying on `C++` operator overloading of `=`, that is probably an edge case. We can document this as a feature, but not make it standard. I don't want to lose performance in C++ unless users specify this option.

File: sort.h [hunk 3]
```diff
@@ -797,6 +807,34 @@ void BITONIC_SORT(SORT_TYPE *dst, const size_t size) {
...
+  for (size_t i = 0; i < size; ++i) {
+    dst[i] = src[i];
+  }
+}
+
+void SORT_TYPE_MOVE(SORT_TYPE *dst, SORT_TYPE *src, const size_t size) {
+  if (dst < src) {
+    SORT_TYPE_CPY(dst, src, size);
+  } else if (dst != src && size > 0) {
+    for (size_t i = size - 1; i > 0; --i) {
```
Comment (swenson) [inline sort.h]: Ditto here about declaring variables at the top.

File: sort.h [hunk 4]
```diff
@@ -108,7 +117,7 @@ static int clzll(uint64_t x) {
 static __inline int compute_minrun(const uint64_t size) {
   const int top_bit = 64 - CLZ(size);
   const int shift = MAX(top_bit, 6) - 6;
-  const int minrun = size >> shift;
+  const int minrun = (int)(size >> shift);
```
Comment (swenson) [inline sort.h]: Are all of these casts necessary when are assigning to an `int`?

Also, could you run `make format`? I think the formatting might be off.

Comment (Baobaobear) [reply sort.h]: Not necessary but some compiler like VS give warnings. I have another question, c89 already supported uint64_t and stdint.h?

**Comment by swenson** [comment]
I will take another look at this PR this week. Sorry for the delay.

**Comment by Baobaobear** [comment]
There was conflicting, I create another pr latter

**Pull request closed without merging**